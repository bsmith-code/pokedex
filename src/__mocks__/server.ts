import {
  mockPokemonDetails,
  mockPokemonEvolutions,
  mockPokemonList,
  mockPokemonSpecies
} from '__mocks__/pokemon'
import { http, HttpResponse } from 'msw'
import { setupServer } from 'msw/node'

export const prepareRoute = (route: string) =>
  `${process.env.REACT_APP_API_BASE_URL ?? ''}/${route}`

const handlers = [
  http.get(prepareRoute('pokemon'), () => HttpResponse.json(mockPokemonList)),
  http.get(prepareRoute('pokemon/:name'), () =>
    HttpResponse.json(mockPokemonDetails)
  ),
  http.get(prepareRoute('pokemon-species/:name'), () =>
    HttpResponse.json(mockPokemonSpecies)
  ),
  http.get(prepareRoute('evolution-chain/:chainId'), () =>
    HttpResponse.json(mockPokemonEvolutions)
  )
]

export const mockServer = setupServer(...handlers)
