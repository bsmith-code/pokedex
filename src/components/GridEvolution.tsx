import { IEvolutionChain } from 'pokeapi-typescript'

import { Box, Grid, ListItemButton, styled, Typography } from '@mui/material'

import { usePokemonContext } from 'context/PokemonContext'

import { flattenChain, getIdFromUrl, getPokemonSprite } from 'utils'

interface IProps {
  isFetching: boolean
  evolutions?: IEvolutionChain
}

const StyledGridItem = styled(ListItemButton)(({ theme }) => ({
  flexDirection: 'column',
  borderRadius: '4px',
  color: '#fff',
  padding: theme.spacing(4),
  backgroundColor: theme.palette.secondary.main,
  '&:hover': {
    backgroundColor: '#6704c9'
  }
}))

export const GridEvolution = ({ evolutions, isFetching }: IProps) => {
  const { handleChangePokemon } = usePokemonContext()
  const flattenedChain = flattenChain(evolutions?.chain)

  return !isFetching ? (
    <Box>
      <Typography variant="h6" sx={{ mb: 2 }}>
        Evolutions
      </Typography>
      {flattenedChain?.length ? (
        <Grid container spacing={2}>
          {flattenedChain.map(({ url, name }) => {
            const id = getIdFromUrl(url)
            return (
              <Grid item xs={12} sm={6} md={6} lg={4} key={`evolution-${name}`}>
                <StyledGridItem onClick={() => handleChangePokemon(name)}>
                  <img
                    alt={name}
                    src={getPokemonSprite(id)}
                    width="120px"
                    height="120px"
                    style={{ objectFit: 'contain' }}
                  />
                  <Typography variant="subtitle2">{name}</Typography>
                </StyledGridItem>
              </Grid>
            )
          })}
        </Grid>
      ) : (
        <Typography>No available evolutions</Typography>
      )}
    </Box>
  ) : (
    <Typography>Loading...</Typography>
  )
}
