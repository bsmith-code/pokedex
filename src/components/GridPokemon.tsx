import {
  Box,
  Grid,
  ListItemButton,
  Pagination,
  styled,
  Typography
} from '@mui/material'

import { usePokemonContext } from 'context/PokemonContext'

import { getIdFromUrl, getPokemonSprite } from 'utils'

import {
  TESTID_GRID_ERROR,
  TESTID_GRID_LOADING,
  TESTID_GRID_POKEMON
} from 'constants/testIds'

const StyledGridItem = styled(ListItemButton)`
  flex-direction: column;
  border-radius: 4px;
  border: 1px solid #efefef;
`

export const GridPokemon = () => {
  const {
    pokemon,
    currentPage,
    paginatedPokemon,
    isErrorPokemon,
    isFetchingPokemon,
    handleChangePage,
    handleChangePokemon
  } = usePokemonContext()

  if (isFetchingPokemon) {
    return <div data-testid={TESTID_GRID_LOADING}>Loading...</div>
  }

  if (!pokemon.length || isErrorPokemon) {
    return <div data-testid={TESTID_GRID_ERROR}>Pokemon unavailable.</div>
  }

  return (
    <Box mt={4} data-testid={TESTID_GRID_POKEMON}>
      <Grid container spacing={2}>
        {paginatedPokemon.map(({ url, name }) => {
          const imageSrc = getPokemonSprite(getIdFromUrl(url))

          return (
            <Grid
              item
              xs={12}
              sm={6}
              md={4}
              lg={3}
              key={`pokemon-grid-${name}`}
            >
              <StyledGridItem onClick={() => handleChangePokemon(name)}>
                <img
                  alt={name}
                  src={imageSrc}
                  width="120px"
                  height="120px"
                  style={{ objectFit: 'contain' }}
                />
                <Typography>{name}</Typography>
              </StyledGridItem>
            </Grid>
          )
        })}
      </Grid>
      <Box mt={6} display="flex" justifyContent="center">
        <Pagination
          page={currentPage}
          count={Math.floor(pokemon.length / 20)}
          onChange={(_, page) => handleChangePage(page)}
        />
      </Box>
    </Box>
  )
}
