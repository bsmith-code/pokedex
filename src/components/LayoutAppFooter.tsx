import { Toolbar, Typography } from '@mui/material'

const LayoutAppFooter = () => {
  const year = new Date().getFullYear()

  return (
    <Toolbar
      variant="dense"
      component="footer"
      sx={{ bgcolor: 'primary.main', justifyContent: 'center' }}
    >
      <Typography color="#fff">
        &copy; Brian Matthew Smith, {year}. All Rights Reserved.
      </Typography>
    </Toolbar>
  )
}

export default LayoutAppFooter
