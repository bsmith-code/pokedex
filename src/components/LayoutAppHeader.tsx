import { useNavigate } from 'react-router-dom'

import { IconButton, Toolbar } from '@mui/material'

const LayoutAppHeader = () => {
  const navigate = useNavigate()

  const handleLogoClick = () => {
    navigate('/')
  }

  return (
    <Toolbar component="header" sx={{ bgcolor: 'primary.main' }}>
      <IconButton onClick={handleLogoClick}>
        <img
          alt="PrizePicks"
          src="/assets/logo.svg"
          width="140px"
          height="auto"
        />
      </IconButton>
    </Toolbar>
  )
}

export default LayoutAppHeader
