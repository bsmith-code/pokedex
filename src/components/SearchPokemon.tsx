import { INamedApiResource } from 'pokeapi-typescript'

import {
  Autocomplete,
  createFilterOptions,
  ListItemButton,
  TextField
} from '@mui/material'

import { useChangePokemon } from 'hooks/useChangePokemon'

interface IProps {
  pokemon: INamedApiResource<void>[]
}

export const SearchPokemon = ({ pokemon }: IProps) => {
  const { handleChangePokemon } = useChangePokemon()

  return (
    <Autocomplete
      filterOptions={createFilterOptions({ limit: 20 })}
      onChange={(_, value) => handleChangePokemon(value?.name ?? '')}
      options={pokemon}
      renderInput={params => <TextField {...params} label="Search Pokemon" />}
      getOptionLabel={({ name }) => name}
      renderOption={(props, { name }) => (
        <ListItemButton
          {...props}
          component="li"
          key={`pokemon-list-${name}`}
          data-testid={`pokemon-${name}`}
        >
          {name}
        </ListItemButton>
      )}
    />
  )
}
