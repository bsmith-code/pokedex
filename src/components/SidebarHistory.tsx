import { Box, List, ListItemButton, Paper, Typography } from '@mui/material'

import { usePokemonContext } from 'context/PokemonContext'

import { TESTID_SEARCH_EMPTY, TESTID_SEARCH_HISTORY } from 'constants/testIds'

export const SidebarHistory = () => {
  const { searchHistory, handleChangePokemon } = usePokemonContext()

  return (
    <Box
      px={4}
      py={3}
      flexShrink={1}
      flexBasis={300}
      component={Paper}
      bgcolor="grey.200"
      sx={{ '@media (max-width: 767px)': { order: 2, flexGrow: 1 } }}
    >
      <Typography variant="subtitle2">Search History</Typography>

      {searchHistory.length ? (
        <List data-testid={TESTID_SEARCH_HISTORY}>
          {searchHistory.map(name => (
            <ListItemButton
              component="li"
              key={`search-${name}`}
              onClick={() => handleChangePokemon(name)}
            >
              {name}
            </ListItemButton>
          ))}
        </List>
      ) : (
        <Typography data-testid={TESTID_SEARCH_EMPTY} sx={{ mt: 2 }}>
          No previous searches.
        </Typography>
      )}
    </Box>
  )
}
