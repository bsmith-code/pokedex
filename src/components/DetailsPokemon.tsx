import { useNavigate } from 'react-router-dom'

import ArrowBackIcon from '@mui/icons-material/ArrowBack'
import {
  Box,
  Card,
  CardContent,
  CardHeader,
  IconButton,
  Typography
} from '@mui/material'

import { usePokemonDetails } from 'hooks/usePokemonDetails'

import { GridEvolution } from 'components/GridEvolution'

import {
  TESTID_DETAILS_ERROR,
  TESTID_DETAILS_LOADING,
  TESTID_DETAILS_POKEMON,
  TESTID_DETAILS_RETURN
} from 'constants/testIds'

export const DetailsPokemon = () => {
  const {
    name,
    details,
    sprites,
    evolutions,
    speciesName,
    isErrorDetails,
    isFetchingDetails,
    isFetchingEvolutions
  } = usePokemonDetails()
  const navigate = useNavigate()

  const handleGoBack = () => {
    navigate('/')
  }

  if (isFetchingDetails) {
    return <div data-testid={TESTID_DETAILS_LOADING}>Loading...</div>
  }

  if (isErrorDetails) {
    return <div data-testid={TESTID_DETAILS_ERROR}>Invalid Pokemon.</div>
  }

  return (
    <>
      <Box display="flex" mt={2} data-testid={TESTID_DETAILS_POKEMON}>
        <Box flexShrink={0} position="relative">
          <IconButton
            onClick={handleGoBack}
            data-testid={TESTID_DETAILS_RETURN}
            sx={{ position: 'absolute', top: 0, left: 0 }}
          >
            <ArrowBackIcon />
          </IconButton>
          <img
            alt={name}
            src={sprites?.other?.['official-artwork']?.front_default ?? ''}
          />
        </Box>
        <Box p={2} flexGrow={1}>
          <Box mb={2}>
            <Typography component="h1" variant="h4">
              {name}
            </Typography>
            <Typography fontSize={14} color="grey.700" sx={{ mt: 1 }}>
              species: {speciesName}
            </Typography>
          </Box>
          {details.map(({ title, content }) => (
            <Card
              key={`details-${title}`}
              sx={{ bgcolor: 'secondary.main', color: 'common.white', mb: 2 }}
            >
              <CardHeader title={title} />
              <CardContent>{content}</CardContent>
            </Card>
          ))}
        </Box>
      </Box>
      <GridEvolution
        evolutions={evolutions}
        isFetching={isFetchingEvolutions}
      />
    </>
  )
}
