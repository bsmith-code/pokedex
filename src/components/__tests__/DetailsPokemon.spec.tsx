import { MemoryRouter, Route, Routes } from 'react-router-dom'
import { mockServer, prepareRoute } from '__mocks__/server'
import { fireEvent, screen, waitFor } from '@testing-library/react'
import { http, HttpResponse } from 'msw'

import { DetailsPokemon } from 'components/DetailsPokemon'

import {
  TESTID_DETAILS_ERROR,
  TESTID_DETAILS_LOADING,
  TESTID_DETAILS_POKEMON,
  TESTID_DETAILS_RETURN
} from 'constants/testIds'

const MockComponent = () => (
  <MemoryRouter initialEntries={['/pokemon/mock-pokemon']}>
    <Routes>
      <Route index path="pokemon?/:pokemonName?" element={<DetailsPokemon />} />
    </Routes>
  </MemoryRouter>
)

describe('DetailsPokemon', () => {
  it('renders correctly while loading', async () => {
    const { container } = customRender(<MockComponent />)

    await waitFor(() => {
      expect(screen.getByTestId(TESTID_DETAILS_LOADING)).toBeInTheDocument()
      expect(container).toMatchSnapshot()
    })
  })

  it('renders correctly with data', async () => {
    const { container } = customRender(<MockComponent />)

    await waitFor(() => {
      expect(screen.getByTestId(TESTID_DETAILS_POKEMON)).toBeInTheDocument()
      expect(container).toMatchSnapshot()
    })
  })

  it('renders correctly with invalid data', async () => {
    mockServer.use(
      http.get(prepareRoute('pokemon/:pokemonName'), () => HttpResponse.error())
    )

    const { container } = customRender(<MockComponent />)

    await waitFor(() => {
      expect(screen.getByTestId(TESTID_DETAILS_ERROR)).toBeInTheDocument()
      expect(container).toMatchSnapshot()
    })
  })

  it('returns to grid correctly', async () => {
    customRender(<MockComponent />)

    const goBackBtn = await screen.findByTestId(TESTID_DETAILS_RETURN)
    fireEvent.click(goBackBtn)

    expect(window.location.pathname).toBe('/')
  })
})
