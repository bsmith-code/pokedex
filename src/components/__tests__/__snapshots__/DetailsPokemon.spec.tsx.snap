// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsPokemon renders correctly while loading 1`] = `
<div>
  <div
    data-testid="details-loading"
  >
    Loading...
  </div>
</div>
`;

exports[`DetailsPokemon renders correctly with data 1`] = `
<div>
  <div
    class="MuiBox-root emotion-0"
    data-testid="details-pokemon"
  >
    <div
      class="MuiBox-root emotion-1"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium emotion-2"
        data-testid="details-return"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-3"
          data-testid="ArrowBackIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
          />
        </svg>
      </button>
      <img
        alt="mock-pokemon #0001"
        src="https://mockPokeApi.com/sprites/pokemon/other/official-artwork/1"
      />
    </div>
    <div
      class="MuiBox-root emotion-4"
    >
      <div
        class="MuiBox-root emotion-5"
      >
        <h1
          class="MuiTypography-root MuiTypography-h4 emotion-6"
        >
          mock-pokemon #0001
        </h1>
        <p
          class="MuiTypography-root MuiTypography-body1 emotion-7"
        >
          species: 
          Mock Species
        </p>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root emotion-8"
      >
        <div
          class="MuiCardHeader-root emotion-9"
        >
          <div
            class="MuiCardHeader-content emotion-10"
          >
            <span
              class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title emotion-11"
            >
              Abilities
            </span>
          </div>
        </div>
        <div
          class="MuiCardContent-root emotion-12"
        >
          mock-ability-0, mock-ability-1, mock-ability-2, mock-ability-3, mock-ability-4, mock-ability-5, mock-ability-6, mock-ability-7, mock-ability-8, mock-ability-9, mock-ability-10, mock-ability-11, mock-ability-12, mock-ability-13, mock-ability-14, mock-ability-15, mock-ability-16, mock-ability-17, mock-ability-18, mock-ability-19
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root emotion-8"
      >
        <div
          class="MuiCardHeader-root emotion-9"
        >
          <div
            class="MuiCardHeader-content emotion-10"
          >
            <span
              class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title emotion-11"
            >
              Moves
            </span>
          </div>
        </div>
        <div
          class="MuiCardContent-root emotion-12"
        >
          mock-move-0, mock-move-1, mock-move-2, mock-move-3, mock-move-4, mock-move-5, mock-move-6, mock-move-7, mock-move-8, mock-move-9, mock-move-10, mock-move-11, mock-move-12, mock-move-13, mock-move-14, mock-move-15, mock-move-16, mock-move-17, mock-move-18, mock-move-19
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root emotion-8"
      >
        <div
          class="MuiCardHeader-root emotion-9"
        >
          <div
            class="MuiCardHeader-content emotion-10"
          >
            <span
              class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title emotion-11"
            >
              Types
            </span>
          </div>
        </div>
        <div
          class="MuiCardContent-root emotion-12"
        >
          mock-type-0, mock-type-1, mock-type-2, mock-type-3, mock-type-4, mock-type-5, mock-type-6, mock-type-7, mock-type-8, mock-type-9, mock-type-10, mock-type-11, mock-type-12, mock-type-13, mock-type-14, mock-type-15, mock-type-16, mock-type-17, mock-type-18, mock-type-19
        </div>
      </div>
    </div>
  </div>
  <p
    class="MuiTypography-root MuiTypography-body1 emotion-23"
    data-testid="grid-loading"
  >
    Loading...
  </p>
</div>
`;

exports[`DetailsPokemon renders correctly with invalid data 1`] = `
<div>
  <div
    data-testid="details-error"
  >
    Invalid Pokemon.
  </div>
</div>
`;
